### PR TITLE
Feature/reminders

### PR DESCRIPTION
**Pull request by kosiakkatrina**
**What**  
Add a scheduled event to send reminders if plans have actions due in 2 days. Schedule once a day for 11am.
image

image

Add a new handler/lambda function to handle the reminders. Change package.json deploy script to separately build the new handler.
Preserve the history of plans being shared in a 'previouslySharedOn' field
**Why**  
To notify customers

**Anything else?**  
 - Have you added any new third-party libraries? node-fetch > because node.js doesn't have fetch by default

### CODE REVIEW

File: router/use-cases/get-reminder-plans.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module.exports = class GetReminderPlans {
+  constructor({ getPlan, planGateway }) {
+    this.getPlan = getPlan;
```
Comment (mmmmillar) [inline router/use-cases/get-reminder-plans.js:3]: This isn't used?

File: router/use-cases/get-plan.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = class GetPlan {
```
Comment (mmmmillar) [inline router/use-cases/get-plan.js:1]: I don't think we need this use case

File: lib/domain/token.js [hunk 3]
```diff
@@ -1,10 +1,11 @@
 import ArgumentError from './argument-error';
 
 export default class Token {
-  constructor({ token, sharedDate }) {
+  constructor({ token, sharedDate, previouslySharedOn }) {
     if (!token) throw new ArgumentError('no token');
     this.token = token;
     this.sharedDate = sharedDate || null;
+    this.previouslySharedOn = previouslySharedOn || [];
```
Comment (mmmmillar) [inline lib/domain/token.js:8]: I'm not sure we need this? I don't know if I'd say sending an auto-reminder is the same as sharing a plan

File: lib/gateways/plan-gateway.js [hunk 4]
```diff
@@ -8,9 +8,10 @@ import {
 } from './models';
 
 class PlanGateway {
-  constructor({ client, tableName }) {
+  constructor({ client, tableName, logger }) {
     this.client = client;
     this.tableName = tableName;
+    this.logger = logger;
```
Comment (mmmmillar) [inline lib/gateways/plan-gateway.js:14]: We're adding a logger but not using it?

File: package.json [hunk 5]
```diff
@@ -4,8 +4,10 @@
   "scripts": {
     "dev": "next dev",
     "build": "next build",
+    "build-handlers": "uniformly build -i router/ -o build/router/router",
+    "postbuild-handlers": "uniformly build -i lib/ -o build/router/lib",
```
Comment (t-davies) [inline package.json:8]: We don't need these any more, since we don't need to transpile now?

File: package.json [hunk 6]
```diff
@@ -40,6 +43,7 @@
     "@babel/plugin-transform-runtime": "^7.9.6",
     "@babel/preset-react": "^7.9.4",
     "@babel/runtime": "^7.9.6",
+    "@nuel/uniformly": "^3.0.4",
```
Comment (t-davies) [inline package.json:46]: Same here, we can remove this if we don't need to transpile.

File: router/use-cases/send-reminder.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        const message = `This is a reminder that your Shared Plan from Hackney Council has actions due in 2 days: ${customerPlanUrl}`;
+
+        await this.smsGateway.sendMessage({
+          name: `Shared Plan Reminder`,
+          number: existingPlan.numbers[0],
+          message,
+          authHeader
+        });
+
+        sortedTokens[0].previouslySharedOn = sortedTokens[0].previouslySharedOn
```
Comment (mmmmillar) [inline router/use-cases/send-reminder.js:29]: I think we can simplify this quite a lot by not worrying about the sharedDate

File: serverless.yml [hunk 8]
```diff
@@ -52,6 +52,25 @@ functions:
       NEXT_PUBLIC_API_URL: https://${self:custom.aliases.${self:provider.stage}}/api
       NEXT_PUBLIC_URL: https://${self:custom.aliases.${self:provider.stage}}
 
+  hn-shared-plan-reminders:
+    name: ${self:service}-reminders-${self:provider.stage}
+    handler: build/router/router/cloudwatch/reminders.handler
```
Comment (t-davies) [inline serverless.yml:57]: This can just be `router/cloudwatch/reminders.handler` now I think?

File: serverless.yml [hunk 9]
```diff
@@ -52,6 +52,25 @@ functions:
...
+    events:
+      - schedule: cron(0 12 * * ? *) # < every day at 12, rate(1 minute) > every minute
+
+    environment:
+      ENV: ${self:provider.stage}
+      ALLOWED_GROUPS: 'housingneeds-singleview-beta'
+      JWT_SECRET: ${ssm:/common/hackney-jwt-secret}
+      SMS_API_URL: ${ssm:/shared-plan/SMS_API_URL}
+      PLANS_TABLE_NAME: ${self:provider.environment.PLANS_TABLE}
+      NEXT_PUBLIC_API_URL: https://${self:custom.aliases.${self:provider.stage}}/api
```
Comment (mmmmillar) [inline serverless.yml:71]: do we use this env var?

File: lib/domain/token.js (hunk 3 above)
Comment (kosiakkatrina) [reply lib/domain/token.js:8]: hm, I see, Chris W mentioned the ideally we will preserve the history of when the plan was first shared.
I agree that this is a little bit 'meh', but do we display an old date in 'last shared with resident' section then?

File: lib/gateways/plan-gateway.js (hunk 4 above)
Comment (kosiakkatrina) [reply lib/gateways/plan-gateway.js:14]: Good point. It was used in a method that I ended up extracting. Will fix

File: serverless.yml (hunk 8 above)
Comment (kosiakkatrina) [reply serverless.yml:57]: Hm, so I tried that and it didn't seem to work, but I'll try again

File: lib/domain/token.js (hunk 3 above)
Comment (mmmmillar) [reply lib/domain/token.js:8]: an auto-reminder isn't the same as an officer sharing the plan with the resident

Comment (mmmmillar) [reply lib/domain/token.js:8]: so, we dont set the shared field at all when sending reminder

Comment (mmmmillar) [reply lib/domain/token.js:8]: and we just send the same token as last time

File: serverless.yml (hunk 9 above)
Comment (kosiakkatrina) [reply serverless.yml:71]: nope

File: lib/domain/token.js (hunk 3 above)
Comment (kosiakkatrina) [reply lib/domain/token.js:8]: 👌

File: serverless.yml (hunk 8 above)
Comment (mmmmillar) [reply serverless.yml:57]: Yeah this shouldn't reference the build folder - ```router/cloudwatch/reminders.handler``` should be right

Comment (kosiakkatrina) [reply serverless.yml]: 👌

**Pull request merged**